src/lib/operators/table_scan.cpp

```diff
@@ -0,0 +1,198 @@
+            for (auto index = ValueID(0); index < size; index++) {
+              auto attribute = typed_dict_segment->attribute_vector()->get(i);
+              if (attribute != value) {
+                posList.emplace_back(RowID({i, ChunkOffset(index)}));
+              }
+            }
+            break;
+          }
+          case ScanType::OpLessThan: {
+            auto value = typed_dict_segment->find_in_dict(search_value());
+            for (auto index = ValueID(0); index < size; index++) {
+              auto attribute = typed_dict_segment->attribute_vector()->get(i);
+              if (attribute < value) {
+                posList.emplace_back(RowID({i, ChunkOffset(index)}));
+              }
+            }
+            break;
+          }
+          case ScanType::OpLessThanEquals: {
+            auto value = typed_dict_segment->find_in_dict(search_value());
+            for (auto index = ValueID(0); index < size; index++) {
+              auto attribute = typed_dict_segment->attribute_vector()->get(i);
+              if (attribute <= value) {
+                posList.emplace_back(RowID({i, ChunkOffset(index)}));
+              }
+            }
+            break;
+          }
+          case ScanType::OpGreaterThan: {
+            auto value = typed_dict_segment->find_in_dict(search_value());
+            for (auto index = ValueID(0); index < size; index++) {
+              auto attribute = typed_dict_segment->attribute_vector()->get(i);
+              if (attribute > value) {
+                posList.emplace_back(RowID({i, ChunkOffset(index)}));
+              }
+            }
+            break;
+          }
+          case ScanType::OpGreaterThanEquals: {
+            auto value = typed_dict_segment->find_in_dict(search_value());
+            for (auto index = ValueID(0); index < size; index++) {
+              auto attribute = typed_dict_segment->attribute_vector()->get(i);
+              if (attribute >= value) {
+                posList.emplace_back(RowID({i, ChunkOffset(index)}));
+              }
+            }
+            break;
+          }
+        }
+
+      const auto typed_ref_segment = std::dynamic_pointer_cast<ReferenceSegment>(segment);
+      if(typed_ref_segment != nullptr) {
+        auto pos_list_size = typed_ref_segment->pos_list()->size();
+
+        for(size_t index = 0; index < pos_list_size; index++) {
+          auto row_id = typed_ref_segment->pos_list()->at(index);
+          auto value = typed_ref_segment
+            ->referenced_table()
+            ->get_chunk(row_id.chunk_id)
+            .get_segment(typed_ref_segment->referenced_column_id())
+            ->operator[](row_id.chunk_offset);
+
+          if(
+              comparator(
+                scan_type(),
+                type_cast<Type>(value),
+                type_cast<Type>(search_value())
+              )
+            ) {
+              posList.emplace_back(RowID{i, ChunkOffset(index)});
+            };
+        }
+      }
+      }
```

Comment:
This fails silently when a different Segment type is implemented. We'd argue that you should at least add a `DebugAssert` here.